nucliadb_node/nucliadb_node/shadow_shards.py

```diff
@@ -0,0 +1,253 @@
+        self.metadata.increment_ops(shard_id)
+        await self.save_metadata()
+
+    @property
+    def loaded(self) -> bool:
+        return self._loaded
+
+    @property
+    def metadata(self) -> ShadowMetadata:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        return self._metadata
+
+    def shard_path(self, shard_id: str) -> str:
+        return f"{self._folder}/{shard_id}"
+
+    async def create(self) -> str:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+
+        # Get a unique shard id
+        shard_id = uuid.uuid4().hex
+        while shard_id in self.shards:
+            shard_id = uuid.uuid4().hex
+
+        async with aiofiles.open(self.shard_path(shard_id), mode="x"):
+            pass
+
+        self.shards.add(shard_id)
+        self.metadata.shards[shard_id] = ShadowShardInfo(shard_id=shard_id)
+        await self.save_metadata()
+        return shard_id
+
+    async def delete(self, shard_id: str) -> None:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+
+        if shard_id not in self.shards:
+            raise ShadowShardNotFound()
+
+        shard_path = self.shard_path(shard_id)
+        await aiofiles.os.remove(shard_path)
+        self.shards.remove(shard_id)
+
+        self.metadata.shards.pop(shard_id)
+        await self.save_metadata()
+
+    def exists(self, shard_id: str) -> bool:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        return shard_id in self.shards
+
+    def encode_operation(self, node_operation: NodeOperation) -> bytes:
+        opcode, payload = node_operation
+        if opcode == OperationCode.SET:
+            payload = payload.SerializeToString()  # type: ignore
+        elif opcode == OperationCode.DELETE:
+            payload = payload.encode()  # type: ignore
+        else:
+            ValueError(f"Unknown opcode: {opcode}")
+        encoded = base64.b64encode(payload).decode()  # type: ignore
+        return (opcode + encoded + "\n").encode()
+
+    def decode_operation(self, encoded: str) -> NodeOperation:
+        opcode = OperationCode(encoded[:4])
+        decoded_payload = base64.b64decode(encoded[4:])
+        if opcode == OperationCode.SET:
+            return (opcode, Resource.FromString(decoded_payload))
+        elif opcode == OperationCode.DELETE:
+            return (opcode, decoded_payload.decode())
+        raise ValueError(f"Unknown opcode: {opcode}")
+
+    async def set_resource(self, brain: Resource, shard_id: str) -> None:
+        if not self.loaded:
+            raise ShadowShardsNotLoaded()
+        if not self.exists(shard_id):
+            raise ShadowShardNotFound()
+
+        encoded = self.encode_operation((OperationCode.SET, brain))
+        shard_path = self.shard_path(shard_id)
+        async with aiofiles.open(shard_path, mode="ab") as f:
+            await f.write(encoded)
+
+        await self.increment_ops(shard_id)
```

Comment:
The incremental counting and the requirement to store it on each message is not useful, as the counting can be done reading the shard_path. Your implementation requires to write twice to dick on each message